src/lib.rs

```diff
@@ -67,34 +69,189 @@ pub mod windows {
                     break;
                 }
             }
-
         }
         es.push('"');
         es.into()
     }
 
-    #[test]
-    fn test_escape() {
-        assert_eq!(escape("--aaa=bbb-ccc".into()), "--aaa=bbb-ccc");
-        assert_eq!(escape("linker=gcc -L/foo -Wl,bar".into()),
-        r#""linker=gcc -L/foo -Wl,bar""#);
-        assert_eq!(escape(r#"--features="default""#.into()),
-        r#""--features=\"default\"""#);
-        assert_eq!(escape(r#"\path\to\my documents\"#.into()),
-        r#""\path\to\my documents\\""#);
-        assert_eq!(escape("".into()), r#""""#);
+    /// Determine if a wide byte needs to be escaped.
+    /// Only tabs, newlines, spaces, and double quotes need to be escaped.
+    /// Example:
+    /// ```
+    /// use shell_escape::windows::needs_escape;
+    ///
+    /// assert!(needs_escape(b'"' as u16));
+    /// assert!(needs_escape(b'\t' as u16));
+    /// assert!(needs_escape(b'\n' as u16));
+    /// assert!(needs_escape(b' ' as u16));
+    /// assert!(!needs_escape(b'\\' as u16));
+    /// ```
+    pub fn needs_escape(wide_byte: u16) -> bool {
+        let (high, low) = ((wide_byte >> 8) as u8, (wide_byte & 0xFF) as u8);
+
+        if high > 0 {
+            // High byte is set, so its not an ASCII character and definitely needs escaping.
+            return true;
+        }
+        matches!(low, b'"' | b'\t' | b'\n' | b' ')
+    }
+
+    /// Escape OsStr for the windows cmd.exe shell.
+    ///
+    /// See [here][msdn] for more information.
+    ///
+    /// Example:
+    /// ```
+    /// use shell_escape::windows::escape_os_str;
+    /// use std::ffi::OsStr;
+    ///
+    /// let s = OsStr::new("foo bar");
+    /// let escaped = escape_os_str(s);
+    /// assert_eq!(escaped, OsStr::new(r#""foo bar""#));
+    /// ```
+    /// [msdn]: http://blogs.msdn.com/b/twistylittlepassagesallalike/archive/2011/04/23/everyone-quotes-arguments-the-wrong-way.aspx
+    pub fn escape_os_str(s: &OsStr) -> Cow<'_, OsStr> {
+        let encoded: Vec<u16> = s.encode_wide().collect();
```

Comment:
```suggestion
        let encoded: Vec<u16> = s.encode_wide();
```

Let's avoid this collect and instead clone the iterator.